expression/expression.go

```diff
@@ -1383,6 +1383,11 @@ func PropagateType(evalType types.EvalType, args ...Expression) {
 				newCol.(*Column).RetType = col.RetType.Clone()
 				args[0] = newCol
 			}
+			if col, ok := args[0].(*CorrelatedColumn); ok {
+				newCol := col.Clone()
+				newCol.(*CorrelatedColumn).RetType = col.RetType.Clone()
+				args[0] = newCol
```

Comment:
Repeating the same code twice may be a little ugly. We may need a `SetType()` or `DeepCopy()` method on `Expression` to avoid this.